proposals/0000-backpack.rst

```diff
@@ -0,0 +1,1185 @@
+with ``str-bytestring``, it's unit identifier is
+``concat-indef-0.1-abcdefg[Str=str-bytestring-0.2-xxx:Str]``.
+
+.. _Module:
+
+A **module identifier** is either a concrete module or a module
+variable.  A concrete module consists of a module name (the module
+being identified) and a unit identifier (the library it is a member of.)
+A module variable consists only of a module name, and specifies the
+name of an unfilled requirement.  Instantiation takes place by
+substituting module variables with concrete module identifiers;
+e.g., ``p[A=<A>]:B`` is instantiated to ``p[A=q:C]:B`` by applying
+the substitution ``<A>`` maps to ``q:C``.  We may speak of the
+**free module variables** of a module or unit identifier in the
+conventional sense.  Module variables are bound by the requirements
+of their defining library.  We will use the metavariable ``M`` to
+represent module identifiers, and ``m`` to represent module names.
+
+Module identifiers are used to ascribe unique names to Haskell entities:
+a Haskell entity ``n`` defined in a module with the module identity
+``M`` (notice that the unique name of a declaration in a module is
+not well-specified without extra information, since only module
+names are specified in syntax) is ascribed the unique name ``M.n``:
+two identifiers are only equal if their unique names are equal.
+
+Example: the module identifier for ``Str`` from ``str-bytestring``
+is ``str-bytestring-0.2-xxx:Str``; the module identifier for
+``Concat`` frmo an uninstantiated ``concat-indef`` is
+``concat-indef-0.1-abcdefg[Str=<Str>]:Concat``.
+
+.. _ModuleSubst:
+
+A **module substitution** is a mapping from module names to modules identifiers.
+Any module name occurs only once in a substitution, and a substitution
+is in canonical form if it is sorted by module name.  We will use
+the metavariable ``S`` to represent substitutions.
+
+Module substitutions can be applied to identifiers::
+
+    -- Substitution on UnitId
+    (p[S])⟦S'⟧   = p[S⟦S'⟧]
+
+    -- Substitution on Module
+    <m>⟦⟧        = <m>
+    <m>⟦m=M,  S⟧ = M
+    <m>⟦m'=M, S⟧ = <m>⟦S⟧    (m ≠ m')
+    (P:m)⟦S⟧     = P⟦S⟧:m
+
+    -- Substitution on ModuleSubst (NOT substitution composition)
+    (m=M, S')⟦S⟧ = m=M⟦S⟧, S'⟦S⟧
+
+Mixed libraries
+---------------
+
+Libraries defines a collection of modules and dependencies parametrized
+over a set of required signatures.  Mixed libraries are an *explicit*
+variant of the libraries users write in Cabal files, where the
+instantiation of all dependencies is recorded explicitly via unit
+identities (similar to ML's *applicative functors*).  Mixed libraries
+are the output of mix-in linking performed by Cabal.
+
+Although users are not expected to ever write a mixed library directly,
+mixed libraries serve as an important intermediate language between
+GHC and Cabal: a mixed library represents a set of command
+line flags and input files that can be directly typechecked by GHC.
+
+Mixed library structure
+~~~~~~~~~~~~~~
+
```

Comment:
Say that we're going to typecheck, and that we're going to instantiate it. To steps.